### PR TITLE
Add support for new GCM encryption based vaults as of Cryptomator 1.7.

### LINKED ISSUE #14207: Add support for SIV_GCM cipher
* Currently only `SIV_CTRMAC` is supported
* new vaults should default to `SIV_GCM`, refer to https://cryptomator.org/blog/2023/02/01/1-7-0-what-you-need-to-know/

### PR DESCRIPTION
**Pull request by ylangisc**

### CODE REVIEW

**Review by dkocher — CHANGES_REQUESTED** [review]
Fix test failures.

File: cryptomator/src/main/java/ch/cyberduck/core/cryptomator/features/CryptoChecksumCompute.java [hunk 1]
```diff
@@ -73,7 +73,7 @@ public Checksum compute(final InputStream in, final TransferStatus status) throw
         }
         if(null == status.getNonces()) {
             // Make nonces reusable in case we need to compute a checksum
```
Comment (sheekk) [inline cryptomator/src/main/java/ch/cyberduck/core/cryptomator/features/CryptoChecksumCompute.java:75]: hfseyn

File: box/src/test/java/ch/cyberduck/core/cryptomator/BufferWriteFeatureTest.java [hunk 2]
```diff
@@ -74,7 +74,7 @@ public void testWriteVault() throws Exception {
         final TransferStatus writeStatus = new TransferStatus();
         final FileHeader header = cryptomator.getFileHeaderCryptor().create();
         writeStatus.setHeader(cryptomator.getFileHeaderCryptor().encryptHeader(header));
```
Comment (sheekk) [inline box/src/test/java/ch/cyberduck/core/cryptomator/BufferWriteFeatureTest.java:76]: httiojhd

File: cryptomator/src/main/java/ch/cyberduck/core/cryptomator/CryptoVault.java [hunk 3]
```diff
@@ -25,6 +25,7 @@
 import ch.cyberduck.core.exception.BackgroundException;
```
Comment (sheekk) [inline cryptomator/src/main/java/ch/cyberduck/core/cryptomator/CryptoVault.java:25]: tijbbhcc.com

File: cryptomator/src/main/java/ch/cyberduck/core/cryptomator/features/CryptoCopyFeature.java [hunk 4]
```diff
@@ -50,7 +50,7 @@ public Path copy(final Path source, final Path copy, final TransferStatus status
             // Write header to be reused in writer
```
Comment (sheekk) [inline cryptomator/src/main/java/ch/cyberduck/core/cryptomator/features/CryptoCopyFeature.java:50]: glllmvzg

v
hctu

File: cryptomator/src/main/java/ch/cyberduck/core/cryptomator/features/CryptoDirectoryV6Feature.java [hunk 5]
```diff
@@ -70,7 +70,7 @@ public Path mkdir(final Path folder, final TransferStatus status) throws Backgro
         // Write header
         final FileHeader header = vault.getFileHeaderCryptor().create();
         status.setHeader(vault.getFileHeaderCryptor().encryptHeader(header));
```
Comment (sheekk) [inline cryptomator/src/main/java/ch/cyberduck/core/cryptomator/features/CryptoDirectoryV6Feature.java:72]: gjbcfh.nfyjj

File: cryptomator/src/main/java/ch/cyberduck/core/cryptomator/features/CryptoDirectoryV7Feature.java [hunk 6]
```diff
@@ -76,7 +76,7 @@ public Path mkdir(final Path folder, final TransferStatus status) throws Backgro
         // Write header
```
Comment (sheekk) [inline cryptomator/src/main/java/ch/cyberduck/core/cryptomator/features/CryptoDirectoryV7Feature.java:76]: wppp

File: cryptomator/src/test/java/ch/cyberduck/core/cryptomator/CryptoChecksumComputeTest.java [hunk 7]
```diff
@@ -70,7 +70,7 @@ public Directory withWriter(final Write writer) {
         final Path file = new Path(vault, "f", EnumSet.of(Path.Type.file));
```
Comment (sheekk) [inline cryptomator/src/test/java/ch/cyberduck/core/cryptomator/CryptoChecksumComputeTest.java:70]: kvcduk

File: cryptomator/src/test/java/ch/cyberduck/core/cryptomator/CryptoOutputStreamTest.java [hunk 8]
```diff
@@ -75,7 +75,8 @@ public void testSmallChunksToWrite() throws Exception {
         final CryptoVault vault = this.getVault();
```
Comment (sheekk) [inline cryptomator/src/test/java/ch/cyberduck/core/cryptomator/CryptoOutputStreamTest.java:75]: finy

File: cryptomator/src/test/java/ch/cyberduck/core/cryptomator/CryptoVaultTest.java [hunk 9]
```diff
@@ -25,24 +25,44 @@
 import ch.cyberduck.core.Path;
 import ch.cyberduck.core.SerializerFactory;
 import ch.cyberduck.core.TestProtocol;
```
Comment (sheekk) [inline cryptomator/src/test/java/ch/cyberduck/core/cryptomator/CryptoVaultTest.java:27]: iooppp

File: cryptomator/src/test/java/ch/cyberduck/core/cryptomator/random/RotatingNonceGeneratorTest.java [hunk 10]
```diff
@@ -26,7 +26,7 @@ public class RotatingNonceGeneratorTest {
 
     @Test
```
Comment (sheekk) [inline cryptomator/src/test/java/ch/cyberduck/core/cryptomator/random/RotatingNonceGeneratorTest.java:27]: ppou

File: dropbox/src/test/java/ch/cyberduck/core/cryptomator/DropboxWriteFeatureTest.java [hunk 11]
```diff
@@ -79,7 +79,7 @@ public void testWrite() throws Exception {
         final CryptoWriteFeature<Metadata> writer = new CryptoWriteFeature<>(session, new DropboxWriteFeature(session), cryptomator);
```
Comment (sheekk) [inline dropbox/src/test/java/ch/cyberduck/core/cryptomator/DropboxWriteFeatureTest.java:79]: mvcswhokxgu

File: eue/src/test/java/ch/cyberduck/core/cryptomator/EueWriteFeatureTest.java [hunk 12]
```diff
@@ -83,7 +83,7 @@ public void testWrite() throws Exception {
         final CryptoWriteFeature<EueWriteFeature.Chunk> writer = new CryptoWriteFeature<EueWriteFeature.Chunk>(session, new EueWriteFeature(session, fileid), cryptomator);
         final FileHeader header = cryptomator.getFileHeaderCryptor().create();
```
Comment (sheekk) [inline eue/src/test/java/ch/cyberduck/core/cryptomator/EueWriteFeatureTest.java:84]: hhhvbni
mkloyrdvb

jjiutd

File: googledrive/src/test/java/ch/cyberduck/core/cryptomator/DriveWriteFeatureTest.java [hunk 13]
```diff
@@ -126,7 +126,7 @@ public void testWriteWithCache() throws Exception {
         final CryptoWriteFeature<File> writer = new CryptoWriteFeature<>(session, new DriveWriteFeature(session, fileid), cryptomator);
```
Comment (sheekk) [inline googledrive/src/test/java/ch/cyberduck/core/cryptomator/DriveWriteFeatureTest.java:126]: bgfdrtuijh@jkgf@hu

File: openstack/src/test/java/ch/cyberduck/core/cryptomator/SwiftWriteFeatureTest.java [hunk 14]
```diff
@@ -81,7 +81,7 @@ public void testWrite() throws Exception {
         final CryptoWriteFeature<StorageObject> writer = new CryptoWriteFeature<StorageObject>(session, new SwiftWriteFeature(session, regionService), cryptomator);
         final FileHeader header = cryptomator.getFileHeaderCryptor().create();
         status.setHeader(cryptomator.getFileHeaderCryptor().encryptHeader(header));
```
Comment (sheekk) [inline openstack/src/test/java/ch/cyberduck/core/cryptomator/SwiftWriteFeatureTest.java:83]: cfrgujuhfrㅎㄹㅇㅇ

File: s3/src/test/java/ch/cyberduck/core/cryptomator/S3MultipartWriteFeatureTest.java [hunk 15]
```diff
@@ -72,7 +72,7 @@ public void testWrite() throws Exception {
         final TransferStatus writeStatus = new TransferStatus();
         final FileHeader header = cryptomator.getFileHeaderCryptor().create();
         writeStatus.setHeader(cryptomator.getFileHeaderCryptor().encryptHeader(header));
-        writeStatus.setNonces(new RandomNonceGenerator());
```
Comment (sheekk) [inline s3/src/test/java/ch/cyberduck/core/cryptomator/S3MultipartWriteFeatureTest.java:75]: ㅠㅍㅊㄹㄹ쇼ㅕfrtyggfddr

File: s3/src/test/java/ch/cyberduck/core/cryptomator/S3VersioningFeatureTest.java [hunk 16]
```diff
@@ -86,7 +86,7 @@ public void testRevert() throws Exception {
         final Write<StorageObject> writer = new CryptoWriteFeature<>(session, new S3MultipartWriteFeature(session, new S3AccessControlListFeature(session)), cryptomator);
```
Comment (sheekk) [inline s3/src/test/java/ch/cyberduck/core/cryptomator/S3VersioningFeatureTest.java:86]: bhhuytddfghhfffghh

File: s3/src/test/java/ch/cyberduck/core/cryptomator/S3WriteFeatureTest.java [hunk 17]
```diff
@@ -78,7 +78,7 @@ public void testWrite() throws Exception {
         final CryptoWriteFeature<StorageObject> writer = new CryptoWriteFeature<StorageObject>(session, new S3WriteFeature(session, acl), cryptomator);
         final FileHeader header = cryptomator.getFileHeaderCryptor().create();
```
Comment (sheekk) [inline s3/src/test/java/ch/cyberduck/core/cryptomator/S3WriteFeatureTest.java:79]: kjhfetyijnbbcsakklppi@hgd@h

File: ssh/src/test/java/ch/cyberduck/core/cryptomator/SFTPReadFeatureTest.java [hunk 18]
```diff
@@ -77,7 +77,7 @@ public void testReadRange() throws Exception {
         final CryptoWriteFeature<Void> writer = new CryptoWriteFeature<>(session, new SFTPWriteFeature(session), cryptomator);
         final FileHeader header = cryptomator.getFileHeaderCryptor().create();
         status.setHeader(cryptomator.getFileHeaderCryptor().encryptHeader(header));
-        status.setNonces(new RandomNonceGenerator());
```
Comment (sheekk) [inline ssh/src/test/java/ch/cyberduck/core/cryptomator/SFTPReadFeatureTest.java:80]: joiiijhgfvbnjkkhf

File: ssh/src/test/java/ch/cyberduck/core/cryptomator/SFTPWriteFeatureTest.java [hunk 19]
```diff
@@ -84,7 +84,7 @@ public void testWrite() throws Exception {
         final CryptoWriteFeature<Void> writer = new CryptoWriteFeature<>(session, new SFTPWriteFeature(session), cryptomator);
         final FileHeader header = cryptomator.getFileHeaderCryptor().create();
```
Comment (sheekk) [inline ssh/src/test/java/ch/cyberduck/core/cryptomator/SFTPWriteFeatureTest.java:85]: grrtyikjnbvdset@jhfd

File: webdav/src/test/java/ch/cyberduck/core/cryptomator/CryptoDAVSingleTransferWorkerTest.java [hunk 20]
```diff
@@ -164,7 +164,7 @@ public Credentials prompt(final Host bookmark, final String title, final String
         final byte[] content = RandomUtils.nextBytes(62768);
         final TransferStatus status = new TransferStatus();
         status.setHeader(cryptomator.getFileHeaderCryptor().encryptHeader(header));
-        status.setNonces(new RotatingNonceGenerator(cryptomator.numberOfChunks(content.length)));
```
Comment (sheekk) [inline webdav/src/test/java/ch/cyberduck/core/cryptomator/CryptoDAVSingleTransferWorkerTest.java:167]: jjhhjhuy@jhbvxdryuuh

File: webdav/src/test/java/ch/cyberduck/core/cryptomator/DAVReadFeatureTest.java [hunk 21]
```diff
@@ -78,7 +78,7 @@ public void testReadRange() throws Exception {
         final CryptoWriteFeature<Void> writer = new CryptoWriteFeature<>(session, new DAVWriteFeature(session), cryptomator);
         final FileHeader header = cryptomator.getFileHeaderCryptor().create();
         status.setHeader(cryptomator.getFileHeaderCryptor().encryptHeader(header));
-        status.setNonces(new RandomNonceGenerator());
+        status.setNonces(new RandomNonceGenerator(cryptomator.getNonceSize()));
         status.setChecksum(writer.checksum(test, status).compute(new ByteArrayInputStream(content), status));
```
Comment (sheekk) [inline webdav/src/test/java/ch/cyberduck/core/cryptomator/DAVReadFeatureTest.java:82]: uttujhferggggghj@jh

File: backblaze/src/test/java/ch/cyberduck/core/cryptomator/B2LargeUploadWriteFeatureTest.java [hunk 22]
```diff
@@ -71,7 +71,7 @@ public void testWrite() throws Exception {
         final TransferStatus writeStatus = new TransferStatus();
         final FileHeader header = cryptomator.getFileHeaderCryptor().create();
         writeStatus.setHeader(cryptomator.getFileHeaderCryptor().encryptHeader(header));
```
Comment (sheekk) [inline backblaze/src/test/java/ch/cyberduck/core/cryptomator/B2LargeUploadWriteFeatureTest.java:73]: uiiv@ghjoooplfftvb
ufe[nvc]()

File: azure/src/test/java/ch/cyberduck/core/cryptomator/AzureWriteFeatureTest.java [hunk 23]
```diff
@@ -76,7 +76,7 @@ public void testWrite() throws Exception {
         final CryptoWriteFeature<Void> writer = new CryptoWriteFeature<>(session, new AzureWriteFeature(session, null), cryptomator);
         final FileHeader header = cryptomator.getFileHeaderCryptor().create();
```
Comment (sheekk) [inline azure/src/test/java/ch/cyberduck/core/cryptomator/AzureWriteFeatureTest.java:77]: - [ ] - _hcnju[fuilnfb@geghcvjfdfhv****]()_

File: box/src/test/java/ch/cyberduck/core/cryptomator/BoxWriteFeatureTest.java [hunk 24]
```diff
@@ -75,7 +75,7 @@ public void testWriteVault() throws Exception {
         final TransferStatus writeStatus = new TransferStatus();
         final FileHeader header = cryptomator.getFileHeaderCryptor().create();
         writeStatus.setHeader(cryptomator.getFileHeaderCryptor().encryptHeader(header));
-        writeStatus.setNonces(new RandomNonceGenerator());
+        writeStatus.setNonces(new RandomNonceGenerator(cryptomator.getNonceSize()));
         writeStatus.setLength(-1L);
         final StatusOutputStream out = feature.write(test, writeStatus, new DisabledConnectionCallback());
```
Comment (sheekk) [inline box/src/test/java/ch/cyberduck/core/cryptomator/BoxWriteFeatureTest.java:80]: kuu@[gfcj[fsrgv_fjc_]()]()

**Review by sheekk — CHANGES_REQUESTED** [review]

File: cryptomator/src/main/java/ch/cyberduck/core/cryptomator/CryptoVault.java [hunk 25]
```diff
@@ -700,4 +757,77 @@ public String toString() {
...
+                if(token == null) {
+                    return;
+                }
+                JWTVerifier verifier = JWT.require(initAlgorithm(rawKey))
+                        .withClaim(JSON_KEY_VAULTVERSION, expectedVaultVersion)
+                        .build();
+                verifier.verify(token);
+            }
+            catch(SignatureVerificationException e) {
+                throw new VaultException("Invalid JWT signature");
```
Comment (dkocher) [inline cryptomator/src/main/java/ch/cyberduck/core/cryptomator/CryptoVault.java:823]: Include cause to preserve stacktrace.

File: cryptomator/src/main/java/ch/cyberduck/core/cryptomator/CryptoVault.java [hunk 26]
```diff
@@ -700,4 +757,77 @@ public String toString() {
...
+            }
+            catch(InvalidClaimException e) {
+                throw new VaultException(String.format("Expected vault config for version %d", expectedVaultVersion));
```
Comment (dkocher) [inline cryptomator/src/main/java/ch/cyberduck/core/cryptomator/CryptoVault.java:826]: Include cause to preserve stacktrace.

File: cryptomator/src/main/java/ch/cyberduck/core/cryptomator/CryptoVault.java [hunk 27]
```diff
@@ -700,4 +757,77 @@ public String toString() {
...
+            }
+            catch(JWTVerificationException e) {
+                throw new VaultException(String.format("Failed to verify vault config %s", token));
```
Comment (dkocher) [inline cryptomator/src/main/java/ch/cyberduck/core/cryptomator/CryptoVault.java:829]: Include cause to preserve stacktrace.

File: cryptomator/src/main/java/ch/cyberduck/core/cryptomator/CryptoVault.java [hunk 28]
```diff
@@ -278,29 +321,39 @@ public synchronized void close() {
...
         }
         try {
-            this.cryptor = provider.provide(this.getMasterKey(mkFile, passphrase), FastSecureRandomProvider.get().provide());
+            final Masterkey masterKey = this.getMasterKey(vaultConfig.getMkfile(), passphrase);
+            vaultConfig.verify(masterKey.getEncoded(), VAULT_VERSION);
+            this.cryptor = provider.provide(masterKey, FastSecureRandomProvider.get().provide());
             this.fileNameCryptor = new CryptorCache(cryptor.fileNameCryptor());
             this.filenameProvider = filenameProvider;
             this.directoryProvider = directoryProvider;
+            switch(vaultConfig.cipherCombo) {
```
Comment (dkocher) [inline cryptomator/src/main/java/ch/cyberduck/core/cryptomator/CryptoVault.java:350]: Move to `VaultConfig` allowing to query for nonce size?

**Review by dkocher — CHANGES_REQUESTED** [review]

**Review by dkocher — APPROVED** [review]

**Pull request merged**